### PR TITLE
HAI-1538 cycling infra

### PR DESCRIPTION
**Pull request by esaostring**
# Description

Created GIS tormays-materials from cycling routes.

* Combine existing prioritized and main cycling route materials into one
* provide method (directory) for inputting existing materials, not available from public sources

### Jira Issue: 

[HAI-1538](https://helsinkisolutionoffice.atlassian.net/browse/HAI-1538)

## Type of change

- [ ] Bug fix 
- [x] New feature 
- [ ] Other

# Instructions for testing

## Running the code in docker containers.

See README for preparing containers (building containers, copying files). After these initial steps, run fetch and process as separate docker runs. See exact details
* `data/README.md`: for downloading required prerequisite from Sharepoint
* `README.md` : for processing

```
docker-compose up -d gis-db
docker-compose run --rm gis-fetch cycle_infra
docker-compose run --rm gis-process cycle_infra
docker-compose stop gis-db
```

Results are in haitaton-gis-output local directory.

## Testing (Python unittest)

Prerequisite: fetch source materials as per instructions in the README (section: cycle_infra)

Go to directory: `haitaton-backend/scripts/gis-material-update/process`

(create and) activate Python virtual environment in process -directory
run `python -m unittest test/test_cycle_infra.py`

# Checklist:

- [x] I have written new tests (if applicable)
- [x] I have ran the tests myself (if applicable)
- [ ] I have made necessary changes to the documentation, link to confluence
 or other location: 

# Other relevant info

None for now.

[HAI-1538]: https://helsinkisolutionoffice.atlassian.net/browse/HAI-1538?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: scripts/gis-material-update/data/README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+## Cycling infra
+
+Current implementation assumes cycling infra source file in this directory.
+
+Copy [source file](https://helsinginkaupunki.sharepoint.com/:u:/r/sites/KYMPHaitaton/Jaetut%20asiakirjat/General/02_Taustamateriaalit/Laastariaineisto/haitaton_2_input/helsinki_cycleways.gpkg?csf=1&web=1&e=pR3rqm)
+from Sharepoint to this directory.
+
+### Instructions to generate combined file
+
+If (for any reason) there is a need to generate combined cycling infra file,
```
Comment (corvidian) [inline scripts/gis-material-update/data/README.md:22]:
```suggestion
If (for any reason) there is a need to regenerate the cycling infra source file,
```

File: scripts/gis-material-update/data/README.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+- [prioritized routes](https://helsinginkaupunki.sharepoint.com/:u:/r/sites/KYMPHaitaton/Jaetut%20asiakirjat/General/02_Taustamateriaalit/Laastariaineisto/haitaton2spatial/input/priorisoidut_reitit.zip?csf=1&web=1&e=75maG1)
+
+Extract zipped shapefiles.
+
+Create one cycling output file combining both infra data:
+
+```sh
+$ ogr2ogr -nln pp -nlt MULTILINESTRING \
+    -f GPKG helsinki_cycleways.gpkg \
+    -sql "SELECT 'main' as type FROM Helsinki_pp_paareitit" Helsinki_pp_paareitit Helsinki_pp_paareitit
```
[the remark this answers is not in the document]
Comment (corvidian) [reply scripts/gis-material-update/data/README.md:37]: Työhakemiston sisältö:
```
❯ ls -al
total 8960
drwxr-xr-x  13 topias.heinonen  staff      416 Apr 11 15:37 .
drwx------+ 91 topias.heinonen  staff     2912 Apr 11 15:36 ..
-rw-r--r--@  1 topias.heinonen  staff        6 Aug 18  2020 Helsinki_pp_paareitit.cpg
-rw-r--r--@  1 topias.heinonen  staff  2253526 Aug 18  2020 Helsinki_pp_paareitit.dbf
-rw-r--r--@  1 topias.heinonen  staff      381 Aug 18  2020 Helsinki_pp_paareitit.prj
-rw-r--r--@  1 topias.heinonen  staff      604 Aug 18  2020 Helsinki_pp_paareitit.qpj
-rw-r--r--@  1 topias.heinonen  staff  1564808 Aug 18  2020 Helsinki_pp_paareitit.shp
-rw-r--r--@  1 topias.heinonen  staff   115644 Aug 18  2020 Helsinki_pp_paareitit.shx
-rw-r--r--@  1 topias.heinonen  staff   350125 Sep 19  2018 priorisoidut_reitit.dbf
-rw-r--r--@  1 topias.heinonen  staff      381 Sep 19  2018 priorisoidut_reitit.prj
-rw-r--r--@  1 topias.heinonen  staff      604 Sep 19  2018 priorisoidut_reitit.qpj
-rw-r--r--@  1 topias.heinonen  staff   246180 Sep 19  2018 priorisoidut_reitit.shp
-rw-r--r--@  1 topias.heinonen  staff    19396 Sep 19  2018 priorisoidut_reitit.shx
```

[the remark this answers is not in the document]
Comment (esaostring) [reply scripts/gis-material-update/data/README.md:37]: Olitko ladannut nuo kaksi `zip` -tiedostoa ja purkanut ne? Eli hakemisto, jossa ajat `ogr2ogr` sisältää alihakemistoina (sisältöineen):
```
$ ls -l priorisoidut_reitit Helsinki_pp_paareitit
Helsinki_pp_paareitit:
total 3864
-rw-r--r-- 1 esa esa       6 elo    16  2022 Helsinki_pp_paareitit.cpg
-rw-r--r-- 1 esa esa 2253526 elo    16  2022 Helsinki_pp_paareitit.dbf
-rw-r--r-- 1 esa esa     381 elo    16  2022 Helsinki_pp_paareitit.prj
-rw-r--r-- 1 esa esa     604 elo    16  2022 Helsinki_pp_paareitit.qpj
-rw-r--r-- 1 esa esa 1564808 elo    16  2022 Helsinki_pp_paareitit.shp
-rw-r--r-- 1 esa esa  115644 elo    16  2022 Helsinki_pp_paareitit.shx

priorisoidut_reitit:
total 616
-rw-r--r-- 1 esa esa 350125 elo    16  2022 priorisoidut_reitit.dbf
-rw-r--r-- 1 esa esa    381 elo    16  2022 priorisoidut_reitit.prj
-rw-r--r-- 1 esa esa    604 elo    16  2022 priorisoidut_reitit.qpj
-rw-r--r-- 1 esa esa 246180 elo    16  2022 priorisoidut_reitit.shp
-rw-r--r-- 1 esa esa  19396 elo    16  2022 priorisoidut_reitit.shx
```

[the remark this answers is not in the document]
Comment (esaostring) [reply scripts/gis-material-update/data/README.md:37]: Aivan, tuota ohjetta pitää tarkentaa tuolta osin. Jos samassa hakemistossa, pitää viitata tuohon shapefileen näin:
```
ogr2ogr -nln pp -nlt MULTILINESTRING \
    -f GPKG helsinki_cycleways.gpkg \
    -sql "SELECT 'main' as type FROM Helsinki_pp_paareitit" Helsinki_pp_paareitit.shp Helsinki_pp_paareitit
```

[the remark this answers is not in the document]
Comment (esaostring) [reply scripts/gis-material-update/data/README.md:37]: Päivitän ohjeen, hyvä löytö.

File: scripts/gis-material-update/process/modules/cycling_infra.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            raise ValueError("Unknown number of buffer values")
+
+        # buffer lines
+        target_infra_polys = self._process_result_lines.copy()
+        target_infra_polys["geometry"] = target_infra_polys.buffer(buffers[0])
+
+        # save to instance
+        self._process_result_polygons = target_infra_polys
+
+    def persist_to_database(self):
```
Comment (corvidian) [inline scripts/gis-material-update/process/modules/cycling_infra.py:37]: I'm a bit confused. Why do we persist to DB while processing? I get similar-looking output after commenting out the to_postgis calls.

I thought the DB is used during processing in some fashion, but at least here it's not. So why is having a running PostGIS required for running the processing script?

Comment (esaostring) [reply scripts/gis-material-update/process/modules/cycling_infra.py:37]: Good question. All processing results are saved to database after processing, during script run. Initial idea was to enable easier debugging and development (sometimes with richer attribute set than production material versions).

So far need for development in local database has been marginal.

There are at least a couple of scenarios, how to proceed:
* as is. Continue development without any change
* strip database save completely now, including already implemented data processing
* continue as is, but review the need to save materials closer to production time (anticipating a need for debugging produced materials and related algorithms)

**Review by corvidian — APPROVED** [review]

**Pull request merged**